---
title: "Differential_Gene_Analysis"
author: "Skinny Genes"
date: "2024-04-08"
output: html_document
---

```{r setup, include=FALSE}
library(edgeR)
library(limma)
library(ggplot2)
library(tidyr)
library(reshape2)
install.packages("ComplexHeatmap")
install.packages("circlize")
BiocManager::install("ComplexHeatmap", force = TRUE)
install.packages('modeltools')
library(ComplexHeatmap)
library(ComplexHeatmap)
library(circlize)
# Put your own directory that you have you counts.txt file downloaded in 
#setwd("/Users/alexvidal/Desktop/Spring2024Classes/CMDA_Capstone/R_Code/")
setwd("/Users/yelebedesta/CMDA_capstone")
```

## Data Cleaning
```{r}
countData <-read.table(header = TRUE, sep = "", file = "counts.txt")
colnames(countData) <- c("Geneid","Chr", "Start" ,"End", "Strand","Length","0hr1","0hr2","0hr3","Avr1","Avr2","Avr3", "EV1", "EV2","EV3")
sampleInfo <- read.table(header = TRUE, sep = "", file = "counts.txt.summary")
#countData$condition <- substr(rownames(sampleInfo), 1,nchar(rownames(sampleInfo))-1)
groupNames <- c("0hr", "0hr","0hr","AvrRxo1", "AvrRxo1","AvrRxo1","EV","EV","EV")

dge <- DGEList(counts = countData[7:15],group = factor(groupNames))
dim(dge) #84570     9
pseudoCounts <- log2(dge$counts+1)
par(mfrow=c(3,3))
hist(pseudoCounts[, "0hr1"])
hist(pseudoCounts[, "0hr2"])
hist(pseudoCounts[, "0hr3"])
hist(pseudoCounts[, "Avr1"])
hist(pseudoCounts[, "Avr2"])
hist(pseudoCounts[, "Avr3"])
hist(pseudoCounts[, "EV1"])
hist(pseudoCounts[, "EV2"])
hist(pseudoCounts[, "EV3"])
boxplot(pseudoCounts, col="gray")

sort(pseudoCounts[, "Avr1"], decreasing = T)

?sort

dge$counts
#here, we are keeping the genes if it has a count of 100 for at least 2 samples
keep<-rowSums(cpm(dge$counts)>100) >= 2 
dim(dge$counts[keep,]) #3098    9
```

### Real data analysis

```{r}
countData <-read.table(header = TRUE, sep = "", file = "counts.txt")
colnames(countData) <- c("Geneid","Chr", "Start" ,"End", "Strand","Length","0hr1","0hr2","0hr3","Avr1","Avr2","Avr3", "EV1", "EV2","EV3")
groupNames <- c("0hr", "0hr","0hr","AvrRxo1", "AvrRxo1","AvrRxo1","EV","EV","EV")
dge <- DGEList(counts = countData[7:15],group = factor(groupNames))
dge.full <- dge #dimension 84570     9

# this is the sample size before the cpm - dimension of dge is 84570 x 9
before_cpm_cut <- dge.full$samples$lib.size
apply(dge$counts, 2, sum) 
keep<-rowSums(cpm(dge$counts)>10) >= 2
dge$counts <- dge$counts[keep, ]
dim(dge$counts) # 3098    9
dge$samples$lib.size <- colSums(dge$counts)

# this is the sample size after the cpm - dimension of dge is 3098 x 9
after_cpm_cut <- dge$samples$lib.size 

#these are the sample names for our plot
samples <- c("Control 1", "Control 2", "Control 3", "Infected Gene 1", "Infected Gene 2", "Infected Gene 3", "Empty Vector 1", "Empty Vector 2", "Empty Vector 3") 

cpm_before_and_after<- data.frame(samples,before_cpm_cut, after_cpm_cut)
colnames(cpm_before_and_after) <- c("Samples","Before the CPM trim","After the CPM trim")

data_long <- melt(cpm_before_and_after, id.vars = "Samples")

options(repr.plot.width = 25, repr.plot.height = 10) 

ggplot(data_long, aes(x = Samples, y = value, fill = variable)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Before and After the CPM trim",
       x = "Samples",
       y = "Gene Counts",
       fill = "Type")+
  theme_minimal() +  # Change the theme if needed
  theme(
    plot.title = element_text(size = 20),  # Change title font size
    legend.title = element_text(size = 14),  # Change legend title font size
    legend.text = element_text(size = 12)# Change legend text font size
  ) 
```

### Visualization
```{r}
# Why is Avr1 so far from Avr2 and 3
dge <- calcNormFactors(dge)
plotMDS(dge, method = "bcv", col=as.numeric(dge$samples$group), top = 1000, xlab = "Principle Component 1", ylab = "Principle Component 2 ")
?plotPCA
plotPCA(
  dge
)
legend("bottomright", as.character(unique(dge$samples$group)), col=1:3, pch=20)

dge$samples$group

#Try PCA

dc <- prcomp(dge[,c(1:length(dge))]
scores <- dc$scores
autoplot(dc, data = dge, colour = 'blue')

plot(scores[, 1], scores[, 2], 
     xlab = "Principal Component 1", ylab = "Principal Component 2",
     main = "PCA Plot")
```


### Estimating Dispersion
```{r}
dgeDispersion <- estimateCommonDisp(dge, verbose=T)

dgeDispersion <- estimateTagwiseDisp(dgeDispersion)
names(dgeDispersion$span)

plotBCV(dgeDispersion)
```

```{r}
install.packages("factoextra")
library(factoextra)
library(stringr)
install.packages("gtable")
library(gtable)
update.packages(ask = FALSE, checkBuilt = TRUE)
dge_cpm <- cpm(dge, log = TRUE)
res.pca <- prcomp(t(dge_cpm))
colors <- as.factor(str_split(rownames(dge$samples), "_", simplify = TRUE)[,1])
fviz_pca_ind(res.pca,
             col.ind = colors,
             palette = "Set1",
             repel = TRUE,    # Avoid text overlapping
             title = "PCA plot of the 3 samples"
             )

fviz_eig(res.pca, addlabels = TRUE)

fviz_contrib(res.pca, choice = "var", axes = 1, top = 100)
plotMDS(dge)

most_var_genes <- sort(apply(dge_cpm, 1, var), decreasing=TRUE)[1:200]
mydgelist_cpm_most_var <- dge_cpm[most_var_genes,] 
library(pheatmap)
pheatmap(mydgelist_cpm_most_var, show_rownames  = FALSE) 


my_z_score <- function(x){
  return( (x - mean(x)) / sd(x))
}

mydgelist_cpm_most_var2 <-  t(apply(mydgelist_cpm_most_var, 1, my_z_score))
pheatmap(mydgelist_cpm_most_var2, show_rownames  = FALSE)

mydgelist <- calcNormFactors(dge)
mydgelist$samples$norm.factors

data <- data.frame(
  replicate = rep(1:3, each = 3),  # 3 replicates
  "Zerohr" = c(1, 1, 1, 0, 0, 0, 0, 0, 0),  # Values for "0hr" column
  "Avr" = c(0, 0, 0, 1, 1, 1, 0, 0, 0),  # Values for "Avr" column
  "EV" = c(0, 0, 0, 0, 0, 0, 1, 1, 1)  # Values for "EV" column
)


model_matrix <- model.matrix(~ 0 + Zerohr+ Avr + EV, data = data)
design <- model_matrix
design

mydgelist_voomed <- voom(mydgelist,design,plot = TRUE)

mydgelist_fit <- lmFit(mydgelist_voomed)
contrast_matrix <- makeContrasts(Zerohr_vs_Avr = Avr - Zerohr,
                             Zerohr_vs_EV= EV - Zerohr,
                             Avr_vs_EV = EV - Avr,
                             levels=design)
contrast_matrix

mydgelist_fit_contrasts <- contrasts.fit(mydgelist_fit, contrast_matrix)
mydgelist_fit_contrasts<- eBayes(mydgelist_fit_contrasts)
mydgelist_de <- decideTests(mydgelist_fit_contrasts)
summary(mydgelist_de)

vennDiagram(mydgelist_de,
            include=c("up", "down"),
            counts.col=c("red", "blue"),
            circle.col = c("red", "blue", "green3", "orange"))
```


## GLM Estimates of Dispersion
```{r}
design.mat <- model.matrix(~ 0 + dge$samples$group)
colnames(design.mat) <- levels(dge$samples$group)
d2 <- estimateGLMCommonDisp(dge,design.mat)

#what are the functions mean/used for
d2 <- estimateGLMTrendedDisp(d2,design.mat, method="power")
# You can change method to "auto", "bin.spline", "power", "spline", "bin.loess".
# The default is "auto" which chooses "bin.spline" when > 200 tags and "power" otherwise.
d2 <- estimateGLMTagwiseDisp(d2,design.mat)
plotBCV(d2)
```
### Comparing DESeq vs edgeR 
under construction
```{r}
## Install DESeq by doing BiocManager::install("DESeq")
library(DESeq2)
?DESeqDataSetFromMatrix
cds <- DESeqDataSetFromMatrix(countData = countData)
cds <- estimateSizeFactors(cds)
sizeFactors(cds)
```


## DGE
```{r}
et0vsAvr <- exactTest(dgeDispersion, pair=c("0hr","AvrRxo1")) # compare groups 1 and 2
et0hrvsEV <- exactTest(dgeDispersion, pair=c("0hr","EV")) # compare groups 1 and 3
etEVvsAvr <- exactTest(dgeDispersion, pair=c("EV","AvrRxo1")) # compare groups 2 and 3
topTags(etEVvsAvr, n=10)
topTags(et0hrvsEV, n=10) 
topTags(et0vsAvr, n=10)

pval <- .01

# 1 means upregulated -1 meands downregulated 0 is not significant
dg0vsAvr <- decideTestsDGE(et0hrvsEV, adjust.method="BH", p.value=pval)
dgEVvsAvr <- decideTestsDGE(etEVvsAvr, adjust.method="BH", p.value=pval)
dg0vsEV <- decideTestsDGE(et0hrvsEV, adjust.method="BH", p.value=pval)


par(mfrow=c(3, 1))
# WHy is red dots withi blue in 0 vs Avr
de1tags0vsAvr <- rownames(dgeDispersion)[as.logical(dg0vsAvr)] 
plotSmear(et0vsAvr, de.tags=de1tags0vsAvr,xlab = "Average logCPM", main = "Special Gene vs Control")
abline(h = c(-2, 2), col = "blue")
de1tagsEVvsAvr <- rownames(dgeDispersion)[as.logical(dgEVvsAvr)] 
plotSmear(etEVvsAvr, de.tags=de1tagsEVvsAvr,xlab = "Average logCPM", main = "Special Gene vs Empty Vector")
abline(h = c(-2, 2), col = "blue")
de1tags0vsEV <- rownames(dgeDispersion)[as.logical(dg0vsEV)] 
plotSmear(et0hrvsEV, de.tags=de1tags0vsEV,xlab = "Average logCPM", main = "Empty Vector vs Control")
abline(h = c(-2, 2), col = "blue")

```


### GLM Testing for Differential Expression
```{r}

fit <- glmFit(d2, design.mat)
# compare (group 1 - group 2) to 0:
# this is equivalent to comparing group 1 to group 2
lrt0vsAvr <- glmLRT(fit, contrast=c(1,-1,0))
lrt0vsEV <- glmLRT(fit, contrast=c(1,0,-1))
lrtEVvsAvr <- glmLRT(fit, contrast=c(0,1,-1))
topTags(lrt0vsAvr, n=10)
topTags(lrt0vsEV, n=10)
topTags(lrtEVvsAvr, n=10)



# How would you know which gene is over or underexpresed#
pval <- .01
par(mfrow=c(1,3))
de0vsAvr <- decideTestsDGE(lrt0vsAvr, adjust.method="BH", p.value = pval)
de2tags0vsAvr <- rownames(d2)[as.logical(de0vsAvr)]
plotSmear(lrt0vsAvr, de.tags=de2tags0vsAvr, main = "Special Gene vs Control",xlab = "Average logCPM")
abline(h = c(-2, 2), col = "blue")

de0vsEV <- decideTestsDGE(lrt0vsEV, adjust.method="BH", p.value = pval)
de2tags0vsEV <- rownames(d2)[as.logical(de0vsEV)]
plotSmear(lrt0vsEV, de.tags=de2tags0vsEV, main = "Empty Vector vs Control",xlab = "Average logCPM")
abline(h = c(-2, 2), col = "blue")

deEVvsAvr <- decideTestsDGE(lrtEVvsAvr, adjust.method="BH", p.value = pval)
de2tagsEVvsAvr <- rownames(d2)[as.logical(deEVvsAvr)]
plotSmear(lrtEVvsAvr, de.tags=de2tagsEVvsAvr, main = "Special Gene vs Empty Vector",xlab = "Average logCPM")
abline(h = c(-2, 2), col = "blue")
```


### Heatmaps
```{r}
geneSymbol <-read.csv("/Users/yelebedesta/Documents/GitHub/CMDA_Capstone/geneSymbols.csv")
View(geneSymbol)

countData[na.omit()]

zev = as.data.frame(et0hrvsEV$table)
zev$padj = -log10(zev$PValue)

```


```{r}
library(tidyverse) 
library(RColorBrewer)
library(ggrepel)
zev = as.data.frame(et0hrvsEV$table)

zev$padj = -log10(zev$PValue)


top0vsEV = data.frame(topTags(et0hrvsEV, n=20))


zev$diffexpressed <- "NO"
  # if log2Foldchange > 0.6 and pvalue < 0.05, set as "UP"<br /><br /><br />
zev$diffexpressed[zev$logFC > 0.6 & zev$PValue < 0.05] <- "UP"
  # if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"<br /><br /><br />
zev$diffexpressed[zev$logFC < -0.6 & zev$PValue < 0.05] <- "DOWN"
  

zev$index <- rownames(zev)
zev$delabel <- ifelse(zev$index %in% head(zev[order(zev$PValue), "index"], 20), zev$index, NA) 

p1 = ggplot(data = zev, aes(x = logFC, y = -log10(PValue), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') +
  geom_point(size = 1) +
  scale_color_manual(values = c("blue", "grey", "red"), 
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  geom_text_repel(max.overlaps = Inf) + ggtitle('Hour 0 compared to Empty Vector') 

p1
```

```{r}
zavr = as.data.frame(et0vsAvr$table)

zavr$padj = -log10(zavr$PValue)


top0vsEV = data.frame(topTags(et0hrvsEV, n=20))


zavr$diffexpressed <- "NO"
  # if log2Foldchange > 0.6 and pvalue < 0.05, set as "UP"<br /><br /><br />
zavr$diffexpressed[zavr$logFC > 0.6 & zavr$PValue < 0.05] <- "UP"
  # if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"<br /><br /><br />
zavr$diffexpressed[zavr$logFC < -0.6 & zavr$PValue < 0.05] <- "DOWN"
  

zavr$index <- rownames(zavr)
zavr$delabel <- ifelse(zavr$index %in% head(zavr[order(zavr$PValue), "index"], 20), zavr$index, NA) 

p2 = ggplot(data = zavr, aes(x = logFC, y = -log10(PValue), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') +
  geom_point(size = 1) +
  scale_color_manual(values = c("blue", "grey", "red"), 
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  geom_text_repel(max.overlaps = Inf) + ggtitle('Hour 0 vs. AvrRxo1 Expression') 

p2
```

```{r}
evavr = as.data.frame(etEVvsAvr$table)

evavr$padj = -log10(evavr$PValue)


top0vsEV = data.frame(topTags(etEVvsAvr, n=20))


evavr$diffexpressed <- "NO"
  # if log2Foldchange > 0.6 and pvalue < 0.05, set as "UP"<br /><br /><br />
evavr$diffexpressed[evavr$logFC > 0.6 & evavr$PValue < 0.05] <- "UP"
  # if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"<br /><br /><br />
evavr$diffexpressed[evavr$logFC < -0.6 & evavr$PValue < 0.05] <- "DOWN"
  

evavr$index <- rownames(evavr)
evavr$delabel <- ifelse(evavr$index %in% head(evavr[order(evavr$PValue), "index"], 20), evavr$index, NA) 

p3 = ggplot(data = evavr, aes(x = logFC, y = -log10(PValue), col = diffexpressed, label = delabel)) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') +
  geom_point(size = 1) +
  scale_color_manual(values = c("blue", "grey", "red"), 
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  geom_text_repel(max.overlaps = Inf) + ggtitle('AvrRxo1 vs. Empty vector Expression') 

p3
```

```{r}
evavr = as.data.frame(etEVvsAvr$table)

evavr$padj = -log10(evavr$PValue)


top0vsEV = data.frame(topTags(etEVvsAvr, n=20))

top0vsEV_copy <- top0vsEV
install.packages("pheatmap")
library(pheatmap)
rownames(top0vsEV_copy) <- rownames(top0vsEV_copy)
colnames(top0vsEV_copy) <- c("Control", "Empty vector")

# Create heatmap
pheatmap(top0vsEV_copy, 
          scale = "row",    # Scale rows
          clustering_method = "complete",    # Choose clustering method
          color = colorRampPalette(c("blue", "white", "red"))(100),    # Define colors
          main = "Gene Expression Heatmap",    # Set heatmap title
          fontsize = 8,    # Set font size
          cellwidth = 15,    # Set cell width
          cellheight = 15    # Set cell height
 )
```








